### PR TITLE
Separate datapath shutdown from datapath cleanup

### PR DESCRIPTION
**Pull request by ThadHouse**
If the datapath threads are still running, they can receive data, that can race when we are trying to clean up worker threads. By cleaning up the datapath threads before the worker threads, we avoid this race

### CODE REVIEW

File: src/core/library.c [hunk 1]
```diff
@@ -276,6 +276,14 @@ MsQuicLibraryUninitialize(
         MsQuicLib.UnregisteredSession = NULL;
     }
 
+    //
+    // Ensure all datapath workers are shut down, so new connections won't be 
+    // created while we shut down worker threads.
+    //
+    if (MsQuicLib.Datapath != NULL) {
```
Comment (nibanks) [inline src/core/library.c:283]: I don't believe `MsQuicLib.Datapath` should ever be NULL. You can remove the check.

File: src/platform/datapath_winuser.c [hunk 2]
```diff
@@ -770,19 +770,14 @@ QuicDataPathInitialize(
...
 _IRQL_requires_max_(PASSIVE_LEVEL)
 void
-QuicDataPathUninitialize(
+QuicDataPathShutdownWorkers(
     _In_ QUIC_DATAPATH* Datapath
     )
 {
     if (Datapath == NULL) {
         return;
     }
```
Comment (nibanks) [inline src/platform/datapath_winuser.c:779]: Do we ever actually call with a NULL input? According to the SAL annotation, it's not optional. So I think we can just remove (and not add) all these checks.

File: src/inc/quic_datapath.h [hunk 3]
```diff
@@ -247,8 +247,18 @@ QuicDataPathInitialize(
     _Out_ QUIC_DATAPATH* *NewDatapath
     );
 
+//
+// Shuts down all workers in a QUIC Datapath handle.
+//
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicDataPathShutdownWorkers(
```
Comment (nibanks) [inline src/inc/quic_datapath.h:255]: I think I prefer just `QuicDataPathShutdown` instead.

**Comment by ThadHouse** [comment]
I think we're hitting the extended shutdown time anyway. The unit tests are timing out on CI.

File: src/inc/quic_platform_dispatch.h [hunk 4]
```diff
@@ -100,6 +100,12 @@ void
     _In_ QUIC_DATAPATH* Datapath
     );
 
+typedef
+void
+(*QUIC_DATAPATH_SHUTDOWN_WORKERS)(
```
Comment (nibanks) [inline src/inc/quic_platform_dispatch.h:105]: This needs to be shutdown too.

**Comment by ThadHouse** [comment]
Closed in favor of #458. Will keep branch around for a day or 2.

**Pull request closed without merging**